src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -463,6 +470,124 @@ PRIMARY KEY ({primaryKeys})
+        /// Returns scale recommendation i.e. whether to scale in or out the host application. The recommendation is
+        /// made based on both the latest metrics and the trend of increase or decrease in the count of unprocessed
+        /// changes in the user table. In all of the calculations, it is attempted to keep the number of workers minimum
+        /// while also ensuring that the count of unprocessed changes per worker stays under the maximum limit.
+        /// </summary>
+        /// <param name="workerCount">The current worker count for the host application.</param>
+        /// <param name="metrics">The collection of metrics samples to make the scale decision.</param>
+        /// <returns></returns>
+        private ScaleStatus GetScaleStatusCore(int workerCount, SqlTriggerMetrics[] metrics)
+        {
+            const int minSamplesForScaling = 5;
+            const int maxChangesPerWorker = 1000;
+
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None
+            };
+
+            // Do not make a scale decision unless we have enough samples.
+            if (metrics is null || (metrics.Length < minSamplesForScaling))
+            {
+                this._logger.LogInformation($"Requesting no-scaling: Insufficient metrics for making scale decision for table: '{this._userTable.FullName}'.");
+                return status;
+            }
+
+            // Add worker if the count of unprocessed changes per worker exceeds the maximum limit.
+            long lastUnprocessedChangeCount = metrics.Last().UnprocessedChangeCount;
+            this._logger.LogInformation($"Unprocessed change count: {lastUnprocessedChangeCount}, worker count: {workerCount}, maximum changes per worker: {maxChangesPerWorker}.");
+
+            if (lastUnprocessedChangeCount > workerCount * maxChangesPerWorker)
+            {
+                status.Vote = ScaleVote.ScaleOut;
+                this._logger.LogInformation($"Requesting scale-out: Found too many unprocessed changes for table: '{this._userTable.FullName}' relative to the number of workers.");
+                return status;
+            }
+
+            // Check if there is a continuous increase or decrease in count of unprocessed changes.
+            bool isIncreasing = true;
+            bool isDecreasing = true;
+            for (int index = metrics.Length - minSamplesForScaling; index < metrics.Length - 1; index++)
+            {
+                isIncreasing = isIncreasing && metrics[index].UnprocessedChangeCount < metrics[index + 1].UnprocessedChangeCount;
+                isDecreasing = isDecreasing && (metrics[index].UnprocessedChangeCount == 0 || metrics[index].UnprocessedChangeCount > metrics[index + 1].UnprocessedChangeCount);
+            }
+
+            if (isIncreasing)
+            {
+                // Scale out only if the expected count of unprocessed changes would exceed the combined limit after 30 seconds.
```

Comment:
Sure. The 30 seconds will be time in usual cases, but when scaling is happening, the polls on pending change count will be less frequent and the time difference between last three samples can be as huge as 2 minutes. I will need to complicate the algorithm if we need to add dependency on time when the change-counts were collected. Possibly, only compare the latest count with the first sample that was collected 30-seconds before the last sample.

Reply:
Sounds good - we can probably stick with this simple algorithm for now (especially since that's similar to what other triggers are doing) and then revisit this later if there seems to be a need to have a more complex decision being made.

Reply:
Let me check here. If this does not require more than 3-4 lines of changes, I will want to add it since I found the intervals between sample to be changing arbitrarily while scaling is in action. Other triggers I checked are not having the logic to estimate future scaling. If there is continuous increase, even if a small one, they scale out the application. I personally think the scaling in and out happens much more aggressively there than required.

Reply:
This is done. The logic did not turn out to be complicated.